### PR TITLE
search: hack in new search language

### PR DESCRIPTION
**Pull request by keegancsmith**
We want to experiment with a new query language we cooked up in discussions last week. This is the first attempt of putting those ideas into code. Because this is intentionally experimental, this is done in a way where it is minimally invasive and simple: we convert queries into our older syntax just before we send them off to the server.

This is all hidden behind a feature flag "search-simple" which probably only works if you set it as an override in the client (and not via the normal feature flags). This is fine since we are experimenting.

This only implements the file and repo filters so far. Will follow up with more advanced things soon.

Test Plan: pnpm it and playing around with it via "sg start web-standalone" and adding "feat=search-simple" to the url.

image

Part of https://github.com/sourcegraph/sourcegraph/issues/58799

### CODE REVIEW

File: client/shared/src/search/searchSimple.test.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import { describe, expect, test } from 'vitest'
+
+import { hacksGobQueriesToRegex } from './searchSimple'
+
+describe('hacksGobQueriesToRegex', () => {
+    test('gob corner cases', () => {
+        expect(hacksGobQueriesToRegex('')).toEqual('')
+
+        // TODO should this match nothing? Right the filter is a noop
```
Comment (fkling) [inline client/shared/src/search/searchSimple.test.ts:9]: ?

```suggestion
        // TODO should this match nothing? Right now the filter is a noop
```

**Review by fkling — APPROVED** [review]
Accepting, in the spirit of "moving fast"

**Review by jtibshirani — COMMENTED** [review]
+1 to building a quick PoC, and a sort of translation approach makes sense to me. However, I'm wondering if we can tie this together with our existing work for `search-new-keyword`? Maybe the translation could happen as part of query parsing in frontend? At least it'd be nice to use the same feature flag to make it easy to enable it and get feedback.

**Comment by keegancsmith** [comment]
FYI I updated this PR to use our query tokenizer. Was easier than I expected :) I imagine this might be easier to support than I imagined.

> However, I'm wondering if we can tie this together with our existing work for `search-new-keyword`?

Yes that is the state I am aiming for. Right now it is so hacky/experimental I'd rather not break that. IE this exists to play with the idea for us, not the rest of sg.

It feels like query language changes can rapidly be prototyped in a way where as a team we feel comfortable with what it becomes, then we decide the rest of the company should try it out.

In particular I wanna test it more and make sure things don't break too badly with it on in the UX. IE if autocomplete and query highlighting are broken I wanna make sure doesn't distract from our keyword experiment.

> Maybe the translation could happen as part of query parsing in frontend?

Yes, that is likely the final place for this to happen.

> At least it'd be nice to use the same feature flag to make it easy to enable it and get feedback.

I didn't wanna do it here because I was worried about too many other random things breaking.

**Comment by keegancsmith** [comment]
@fkling mind taking another quick pass at the code and critique it from a TS style perspective? Keen to learn.

File: client/shared/src/search/searchSimple.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                    case 'file':
+                    case 'f': {
+                        return gobToRegex(token.value.value)
+                    }
+
+                    default: {
+                        return token.value.value
+                    }
+                }
+            })()
```
Comment (fkling) [inline client/shared/src/search/searchSimple.ts:51]: Unusual for me to see, but I like the idea (I wish there were more kinds of expressions in JS). In general I avoid introducing unnecessary function calls but it's not a problem either.

**Review by fkling — APPROVED** [review]
Good idea to use the scanner (I should have suggested it). I've been working with all that code to improve the query input and it's nice to see it used more.

Overall there isn't any cray TS stuff going on, looks all good to me!

**Review by jtibshirani — APPROVED** [review]
Ah now I understand it's an extra-temporary feature flag :) Works for me.

**Pull request merged**